cpp/src/copying/contiguous_split.cu

```diff
@@ -1158,122 +1265,752 @@ std::vector<packed_table> contiguous_split(cudf::table_view const& input,
+
+      CUDF_CUDA_TRY(cudaMemcpyAsync(h_offsets.data(),
+                                    offsets.data(),
+                                    sizeof(std::size_t) * offsets.size(),
+                                    cudaMemcpyDefault,
+                                    stream.value()));
+
+      // the next part is working on the CPU, so we want to synchronize here
+      stream.synchronize();
+    }
+
+    std::vector<std::size_t> num_batches_per_iteration;
+    std::vector<std::size_t> size_of_batches_per_iteration;
+    std::vector<std::size_t> accum_size_per_iteration;
+    std::size_t accum_size = 0;
+    {
+      auto current_offset_it = h_offsets.begin();
+      // figure out how many iterations we need, while fitting batches to iterations
+      // with no more than user_buffer_size bytes worth of batches
+      while (current_offset_it != h_offsets.end()) {
+        // next_iteration_it points to the batch right above the boundary (the batch
+        // that didn't fit).
+        auto next_iteration_it =
+          std::lower_bound(current_offset_it,
+                           h_offsets.end(),
+                           // We add the cumulative size + 1 because we want to find what would fit
+                           // within a buffer of user_buffer_size (up to user_buffer_size).
+                           // Since h_offsets is a prefix scan, we add the size we accumulated so
+                           // far so we are looking for the next user_buffer_sized boundary.
+                           user_buffer_size + accum_size + 1);
+
+        // we subtract 1 from the number of batch here because next_iteration_it points
+        // to the batch that didn't fit, so it's one off.
+        auto batches_in_iter = std::distance(current_offset_it, next_iteration_it) - 1;
+
+        // to get the amount of bytes in this iteration we get the prefix scan size
+        // and subtract the cumulative size so far, leaving the bytes belonging to this
+        // iteration
+        auto iter_size_bytes = *(current_offset_it + batches_in_iter) - accum_size;
+        accum_size += iter_size_bytes;
+
+        num_batches_per_iteration.push_back(batches_in_iter);
+        size_of_batches_per_iteration.push_back(iter_size_bytes);
+        accum_size_per_iteration.push_back(accum_size);
+
+        if (next_iteration_it == h_offsets.end()) { break; }
+
+        current_offset_it += batches_in_iter;
+      }
+    }
+
+    // apply changed offset
+    {
+      rmm::device_uvector<std::size_t> d_accum_size_per_iteration(
+        accum_size_per_iteration.size(), stream, temp_mr);
+
+      CUDF_CUDA_TRY(cudaMemcpyAsync(d_accum_size_per_iteration.data(),
+                                    accum_size_per_iteration.data(),
+                                    accum_size_per_iteration.size() * sizeof(std::size_t),
+                                    cudaMemcpyDefault,
+                                    stream.value()));
```

Comment:
You can use a `vector_factories` to build a `uvector` from a `std::vector` which basically combines these 2 steps
```suggestion
      auto d_accum_size_per_iteration = cudf::detail::make_device_uvector_async( accum_size_per_iteration, stream, temp_mr);
```
Reference:
https://github.com/rapidsai/cudf/blob/dbf140ef5070a129b21326068d38c7b2bba12ba5/cpp/include/cudf/detail/utilities/vector_factories.hpp#L94-L97

Reply:
Done: https://github.com/rapidsai/cudf/pull/13260/commits/b47f2c761bbf396dc833e660172ab0a1f40514da